File: 1590-make-sum-divisible-by-p/1590-make-sum-divisible-by-p.py
class Solution:
    def minSubarray(self, nums: List[int], p: int) -> int:
        total = sum(nums)
        target = total % p
        if target == 0:
            return 0

        mp = {0: -1}
        prefix = 0
        res = len(nums)

        for i, x in enumerate(nums):
            prefix = (prefix + x) % p
            need = (prefix - target + p) % p

            if need in mp:
                res = min(res, i - mp[need])

            mp[prefix] = i

        return -1 if res == len(nums) else res

File: 1590-make-sum-divisible-by-p/1590-make-sum-divisible-by-p.cpp
class Solution {
public:
    int minSubarray(vector<int>& nums, int p) {
        long total = 0;
        for (int x : nums) total += x;

        long target = total % p;
        if (target == 0) return 0;

        unordered_map<int, int> mp;
        mp[0] = -1;

        long prefix = 0;
        int res = nums.size();

        for (int i = 0; i < nums.size(); i++) {
            prefix = (prefix + nums[i]) % p;
            int need = (prefix - target + p) % p;

            if (mp.count(need))
                res = min(res, i - mp[need]);

            mp[(int)prefix] = i;
        }

        return res == nums.size() ? -1 : res;
    }
};